lib/logstash/plugin_mixins/ecs_compatibility_support/selector.rb

```diff
@@ -102,13 +136,20 @@ def value_from(defined_choices)
 
             fail(ArgumentError, "at least one choice must be defined") if defined_choices.empty?
 
-            missing = @supported_modes - defined_choices.keys
+            missing = @supported_modes - (defined_choices.keys + @alias_map.keys)
             fail(ArgumentError, "missing one or more required choice definition #{missing}") if missing.any?
 
             unknown = defined_choices.keys - @supported_modes
             fail(ArgumentError, "unknown choices #{unknown}; valid choices are #{@supported_modes}") if unknown.any?
 
-            defined_choices.fetch(@active_mode)
+            # resolve aliases of missing choices
+            active_mode = @active_mode
+            while !defined_choices.include?(active_mode) && @alias_map.include?(active_mode)
```

Comment:
I think we can improve `alias_map` by resolving all intermediate relation, :v8 => :v1 => :v2 => :v3  -> :v8 => :v3, in Selector#initialize. Or may be it is too early to optimise? At this moment, it is hard to guess how long will be the chain.

Reply:
Unfortunately, a hash literal with duplicates in Ruby can't be detected. By the time the object is instantiated for us, it has already been deduplicated (last wins).

Reply:
You raise a very good point. 

Currently, almost all of this validation logic is in `Selector`, and `Selector::State` is documented as private and assumes that it is given valid input by the "only" thing that instantiates it. 

But `Selector::State` is also instantiated by the spec helper, which means it is possible to get bad input into it, including a circular chain. At the very least, I should safeguard this to the length of the alias map (longest possible chain), but perhaps this warrants refactoring the validation to ensure it is shared between BOTH paths to instantiation.

Reply:
Upon further reflection, we do _not_ want to resolve the aliases to their deepest available value at start time, because the selector state should pick the first available value in the chain.

I've addressed this in several ways.

First, we now have specs to ensure that we resolve intermediates:

~~~
            context 'when aliases are given' do
              let(:ecs_supported_modes) { [:disabled, :v1, :v8 => :v2, :v2 => :v1 ] }
              let(:ecs_effective_mode) { :v8 }
              it 'selects a given value' do
                expect(ecs_select[disabled: "nope", v1: "nah", v8: "hooray"]).to eq("hooray")
              end
              it 'resolves the deepest alias of a missing value' do
                expect(ecs_select[disabled: "sad-trombone", v1: "wahoo"]).to eq("wahoo")
              end
              it 'resolves the intermediate alias of a missing value' do
                expect(ecs_select[disabled: "sad-trombone", v1: "oh-no", v2: "nice"]).to eq("nice")
              end
            end
~~~

Next, this library now _only_ initializes a`Selector::State` through a new `Selector#state_for(current)`, ensuring that the components have been pre-validated, and this `Selector::State#value_for` safe-guards against circular references again here anyway since it is trivial to do so.

Reply:
It looks good and the safeguard looks promising. Just one minor suggestion. Could you add a comment to remind us "we do not resolve the aliases to their deepest available value"? 
LGTM 😎